README
Spice agent for Linux
=====================

The spice agent for Linux consists of 2 parts, a daemon spice-vdagentd
and a per session process spice-vdagent. The daemon gets started in
Spice guests through a Sys-V initscript or a systemd unit. The per
session gets automatically started in desktop environments which honor
/etc/xdg/autostart, and under gdm.

The main daemon needs to know which session daemon is in the currently
active session (think switch user functionality) for this console kit
or systemd-logind (compile time option) is used. If no session info is
available only one session agent is allowed.

Features:
* Client mouse mode (no need to grab mouse by client, no mouse lag)
  this is handled by the daemon by feeding mouse events into the kernel
  via uinput. This will only work if the active session is running a
  spice-vdagent process so that its resolution can be determined.
* Support of copy and paste (text and images) between the active session
  and the client. This supports both the primary selection and the clipboard.

All vdagent communications on the guest side run over a single pipe which
gets presented to the guest os as a virtio serial port.

Under windows this virtio serial port has the following name:
\\\\.\\Global\\com.redhat.spice.0

Under Linux this virtio serial port has the following name:
/dev/virtio-ports/com.redhat.spice.0

To enable the virtio serial port you need to pass the following params on
the qemu cmdline:

For qemu < 0.14.0:
-device spicevmc

For qemu >= 0.14.0:
-device virtio-serial-pci,id=virtio-serial0,max_ports=16,bus=pci.0,addr=0x5 \
-chardev spicevmc,name=vdagent,id=vdagent \
-device \
virtserialport,nr=1,bus=virtio-serial0.0,chardev=vdagent,name=com.redhat.spice.0


ChangeLog
spice-vdagent 1.0
-----------------

- new session agent to use gtk+, and friendlier license (simpler
  and more portable code)
- remove guest resolution configuration support, done by driver
- remove file transfer support, deprecated with folder sharing

spice-vdagent-0.15.0
--------------------
* Xspice support
* Release clipboard on client disconnect if owned by client (rhbz#1003977)
* Turn some error messages into debugging messages (rhbz#918310)
* Not having the virtio channel is not an error; instead silently do nothing

spice-vdagent-0.14.0
--------------------
* More multi-monitor and arbritary resolution support bugfixes
* Add support for file transfers from client to guest
* Add support for setups with multiple Screens (multiple qxl devices each
  mapped to their own screen), limitations:
 -Max one monitor per Screen / qxl device
 -All monitors / Screens must have the same resolution
 -No client -> guest resolution syncing
* Add spice-vdagent -X cmdline option, which runtime disables console-kit /
  systemd-logind integration for setups where these are not used
* Add manpages for spice-vdagent and spice-vdagentd

spice-vdagent-0.12.1
--------------------
* Various bugfixes for multi-monitor and arbritary resolution support
* Requires libXrandr >= 1.3, Note 0.12.0 also required this, but did not
  check for it. For older distributions use 0.10.1

spice-vdagent-0.12.0
--------------------
* Full multi-monitor and arbritary resolution support, this requires a new
  enough xorg-x11-drv-qxl driver, as well as a new enough host
* systemd service support, using systemd hardware acivation
* Use syslog for logging, rather then logging to private log files

spice-vdagent-0.10.1
--------------------
* Fix a race condition when opening the vdagent virtio serial port, which
  caused it to get opened / closed in rapid succession a number of times
  on vm boot

spice-vdagent-0.10.0
--------------------
* Add limited support for multiple displays, see README
* Add support for RHEL-5 (and other distributions with a non hotplug
  capable Xorg and/or no console-kit), see README.RHEL-5
* Add support for using libsystemd-logind as session information source
  instead of console-kit

spice-vdagent-0.8.1
-------------------
* In daemon mode the session vdagent now retries connecting to the system
  vdagentd every second, once a connection is made a version check is done,
  if the version differs (which only happens on an upgrade from one version
  to the next) the sesion vdagent re-execs itself (Marc-André Lureau)

spice-vdagent-0.8.0
-------------------
* Add support for copy and paste using the primary selection, to use this
  you need a spice-gtk widget based client and the latest spice-gtk code
  (Marc-André Lureau and Hans de Goede)
* Autotoolized (Christophe Fergeau)
* Allow building without consolekit, for systems which don't have ck, such
  as RHEL-5 (Christophe Fergeau)
* Various small bugfixes (Hans de Goede)

spice-vdagent-0.6.3
-------------------
* Initial release, starting with version nr 0.6.3, to indicate that it
  more or less supports all parts of the cdagent protocol in spice-protocol
  and spice 0.6.3


src/vdagent/utils.c
 /* GIO - GLib Input, Output and Streaming Library
 *
 * Copyright (C) Carl-Anton Ingmarsson 2011 <ca.ingmarsson@gmail.com>
 * Copyright 2010-2014 Red Hat, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General
 * Public License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

#include "utils.h"

#include <syslog.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

GStrv strv_from_data(char *data, gsize len, gssize *pos)
{
    GStrv strv;
    gssize i, n;

    g_return_val_if_fail(len >= 2, NULL);

    for (i = 0, n = 0; i < len; i++) {
        if (data[i])
            continue;

        if (i >= 1 && !data[i-1])
            break;
        n++;
    }

    g_return_val_if_fail(i < len, NULL);

    strv = g_new0(gchar *, n + 1);
    strv[n] = NULL;

    for (i = 0, n = 0; i < len; i++) {
        if (!data[i] && i >= 1 && !data[i-1])
            break;

        if (!strv[n])
            strv[n] = g_strndup(data + i, len - i);

        if (!data[i])
            n++;
    }

    if (pos)
        *pos = i + 1;

    return strv;
}

gchar* str_from_data(char *data, gsize len, gssize *pos)
{
    gchar *str;
    int i;

    g_return_val_if_fail(len >= 1, NULL);

    for (i = 0; i < len; i++)
        if (!data[i])
            break;

    g_return_val_if_fail(i < len, NULL);

    str = g_strdup(data);
    if (pos)
        *pos = strlen(str) + 1;

    return str;
}

static void
log_handler(const gchar *log_domain, GLogLevelFlags log_level,
            const gchar *message, gpointer user_data)
{
    int level;

    /* Note that crit and err are the other way around in syslog */

    switch (G_LOG_LEVEL_MASK & log_level) {
    case G_LOG_LEVEL_ERROR:
        level = LOG_CRIT;
        break;
    case G_LOG_LEVEL_CRITICAL:
        level = LOG_ERR;
        break;
    case G_LOG_LEVEL_WARNING:
        level = LOG_WARNING;
        break;
    case G_LOG_LEVEL_MESSAGE:
        level = LOG_NOTICE;
        break;
    case G_LOG_LEVEL_INFO:
        level = LOG_INFO;
        break;
    case G_LOG_LEVEL_DEBUG:
        level = -1;
        break;
    default:
        level = LOG_ERR;
        break;
    }

    /* Log to syslog first */
    if (level != -1) {
        if (log_domain)
            syslog (level, "%s: %s", log_domain, message);
        else
            syslog (level, "%s", message);
    }

    /* And then to default handler for aborting and stuff like that */
    g_log_default_handler (log_domain, log_level, message, user_data);
}

static void
printerr_handler(const gchar *string)
{
    /* Print to syslog and stderr */
    syslog(LOG_WARNING, "%s", string);
    fprintf(stderr, "%s", string);
}

void
gsyslog_install(const char *domain)
{
    GLogLevelFlags flags = G_LOG_FLAG_FATAL | G_LOG_LEVEL_ERROR |
        G_LOG_LEVEL_CRITICAL | G_LOG_LEVEL_WARNING |
        G_LOG_LEVEL_MESSAGE | G_LOG_LEVEL_INFO;

    openlog (domain, LOG_PID, LOG_AUTH);

    g_log_set_handler (NULL, flags, log_handler, NULL);
    g_log_set_handler ("Glib", flags, log_handler, NULL);
    g_log_set_handler ("Gtk", flags, log_handler, NULL);
    g_log_set_default_handler (log_handler, NULL);
    g_set_printerr_handler (printerr_handler);
}

typedef struct
{
    void         *buffer;
    gsize         count;
    int           io_priority;
    GCancellable *cancellable;
    gsize         bytes_read;
} ReadAllData;

static void
free_read_all_data (ReadAllData *read_data)
{
    if (read_data->cancellable)
        g_object_unref (read_data->cancellable);

    g_slice_free (ReadAllData, read_data);
}

static void
read_all_cb (GObject *source_object, GAsyncResult *res, gpointer user_data)
{
    GInputStream *stream = G_INPUT_STREAM (source_object);
    GSimpleAsyncResult *simple = G_SIMPLE_ASYNC_RESULT (user_data);

    gsize bytes_read;
    GError *err = NULL;
    ReadAllData *read_data;

    bytes_read = g_input_stream_read_finish (stream, res, &err);
    g_debug("bytes read %" G_GSIZE_FORMAT, bytes_read);
    if (bytes_read == -1) {
        g_simple_async_result_take_error (simple, err);
        goto done;
    } else if (bytes_read == 0) {
        g_simple_async_result_set_error (simple, G_IO_ERROR, G_IO_ERROR_CLOSED, "stream closed");
        goto done;
    }

    read_data = g_simple_async_result_get_op_res_gpointer (simple);
    read_data->bytes_read += bytes_read;
    if (read_data->bytes_read < read_data->count) {
        g_input_stream_read_async (stream,
                                   (guint8 *)read_data->buffer + read_data->bytes_read,
                                   read_data->count - read_data->bytes_read, 0,
                                   read_data->cancellable, read_all_cb, simple);
        return;
    }

 done:
    g_simple_async_result_complete (simple);
    g_object_unref (simple);
}

/* See also https://bugzilla.gnome.org/show_bug.cgi?id=679662 */
void
input_stream_read_all_async (GInputStream        *stream,
                             void                *buffer,
                             gsize                count,
                             int                  io_priority,
                             GCancellable        *cancellable,
                             GAsyncReadyCallback  callback,
                             gpointer             user_data)
{
    ReadAllData *read_data;
    GSimpleAsyncResult *simple;

    read_data = g_slice_new0 (ReadAllData);
    read_data->buffer = buffer;
    read_data->count = count;
    read_data->io_priority = io_priority;
    if (cancellable)
        read_data->cancellable = g_object_ref (cancellable);

    simple = g_simple_async_result_new (G_OBJECT (stream), callback, user_data,
                                        input_stream_read_all_async);
    g_simple_async_result_set_op_res_gpointer (simple, read_data,
                                               (GDestroyNotify)free_read_all_data);

    g_debug("reading %" G_GSIZE_FORMAT, count);
    g_input_stream_read_async (stream, buffer, count, io_priority, cancellable,
                               read_all_cb, simple);
}

gboolean
input_stream_read_all_finish (GInputStream *stream,
                              GAsyncResult *res,
                              gsize        *bytes_read,
                              GError      **error)
{
    GSimpleAsyncResult *simple;

    g_return_val_if_fail (g_simple_async_result_is_valid (res, G_OBJECT (stream),
                                                          input_stream_read_all_async),
                          FALSE);

    simple = (GSimpleAsyncResult *)res;

    if (g_simple_async_result_propagate_error (simple, error))
        return FALSE;

    if (bytes_read) {
        ReadAllData *read_data;

        read_data = g_simple_async_result_get_op_res_gpointer (simple);
        *bytes_read = read_data->bytes_read;
    }

    return TRUE;
}


src/vdagent/utils.h
/* GIO - GLib Input, Output and Streaming Library
 *
 * Copyright (C) Carl-Anton Ingmarsson 2011 <ca.ingmarsson@gmail.com>
 * Copyright 2010-2014 Red Hat, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General
 * Public License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

#ifndef UTILS_H_
#define UTILS_H_

#include <gio/gio.h>

G_BEGIN_DECLS

void
gsyslog_install(const gchar *domain);

void
input_stream_read_all_async (GInputStream        *stream,
                             void                *buffer,
                             gsize                count,
                             int                  io_priority,
                             GCancellable        *cancellable,
                             GAsyncReadyCallback  callback,
                             gpointer             user_data);

gboolean
input_stream_read_all_finish (GInputStream *stream,
                              GAsyncResult *res,
                              gsize        *bytes_read,
                              GError      **error);

gchar* str_from_data(char *data, gsize len, gssize *pos);
GStrv strv_from_data(char *data, gsize len, gssize *pos);

G_END_DECLS

#endif


src/vdagent/vdagent.h
/*
 * vdagent session
 * Copyright (C) 2014  Red Hat, Inc.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#ifndef VDAGENT_H_
#define VDAGENT_H_

#include <spice/vd_agent.h>

#include <gtk/gtk.h>
#include <gdk/gdkx.h>
#ifdef G_OS_UNIX
#include <gio/gunixsocketaddress.h>
#endif

G_BEGIN_DECLS

#define SPICE_TYPE_VDAGENT            (spice_vdagent_get_type ())
#define SPICE_VDAGENT(obj)            (G_TYPE_CHECK_INSTANCE_CAST ((obj), SPICE_TYPE_VDAGENT, SpiceVDAgent))
#define SPICE_VDAGENT_CLASS(klass)    (G_TYPE_CHECK_CLASS_CAST ((klass), SPICE_TYPE_VDAGENT, SpiceVDAgentClass))
#define SPICE_IS_VDAGENT(obj)         (G_TYPE_CHECK_INSTANCE_TYPE ((obj), SPICE_TYPE_VDAGENT))
#define SPICE_IS_VDAGENT_CLASS(klass) (G_TYPE_CHECK_CLASS_TYPE ((klass), SPICE_TYPE_VDAGENT))
#define SPICE_VDAGENT_GET_CLASS(obj)  (G_TYPE_INSTANCE_GET_CLASS ((obj), SPICE_TYPE_VDAGENT, SpiceVDAgentClass))

enum {
    VDAGENTD_GUEST_XORG_RESOLUTION,
    VDAGENTD_MONITORS_CONFIG,
    VDAGENTD_CLIPBOARD_GRAB,
    VDAGENTD_CLIPBOARD_REQUEST,
    VDAGENTD_CLIPBOARD_DATA,
    VDAGENTD_CLIPBOARD_RELEASE,
    VDAGENTD_VERSION,
    VDAGENTD_FILE_XFER_START,
    VDAGENTD_FILE_XFER_STATUS,
    VDAGENTD_FILE_XFER_DATA,
    VDAGENTD_CLIENT_DISCONNECTED,

    VDAGENTD_LAST
};

enum {
    OWNER_NONE,
    OWNER_GUEST,
    OWNER_CLIENT
};

typedef struct _VDAgentdHeader {
    guint32 type;
    guint32 arg1;
    guint32 arg2;
    guint32 size;
} VDAgentdHeader;

typedef struct _SpiceVDAgent {
    GObject parent;

    GCancellable *cancellable;
    GSocket *socket;
    GSocketConnectable *connectable;
    GIOStream *connection;
    GQueue *outq;
    gboolean writing;
    gsize pos;

    VDAgentdHeader header;
    gpointer data;

    int clipboard_owner[G_MAXUINT8];
    struct {
        GMainLoop *loop;
        GtkSelectionData *selection_data;
        guint info;
        gint selection;
    } clipboard_get;
} SpiceVDAgent;

typedef struct _SpiceVDAgentClass {
    GObjectClass parent;
} SpiceVDAgentClass;

GType spice_vdagent_get_type(void);

void spice_vdagent_write            (SpiceVDAgent *self,
                                     gpointer data, guint32 size,
                                     GFreeFunc free_func);
void spice_vdagent_write_header     (SpiceVDAgent *self,
                                     guint32 type, guint32 arg1, guint32 arg2, guint32 size);
void spice_vdagent_write_msg        (SpiceVDAgent *self,
                                     guint32 type, guint32 arg1, guint32 arg2,
                                     gpointer data, guint32 size, GFreeFunc free_func);

G_END_DECLS

#endif /* VDAGENT_H_ */


src/vdagent/vdagent-clipboard.h
/*
 * vdagent session
 * Copyright (C) 2014  Red Hat, Inc.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#ifndef VDAGENT_CLIPBOARD_H_
#define VDAGENT_CLIPBOARD_H_

#include <gtk/gtk.h>
#include "vdagent.h"

G_BEGIN_DECLS

void vdagent_clipboard_init             (SpiceVDAgent *agent);
void vdagent_clipboard_request          (SpiceVDAgent *agent, guint8 selection,
                                         const gchar *type);
void vdagent_clipboard_grab             (SpiceVDAgent *agent, guint8 selection,
                                         const GStrv types);
void vdagent_clipboard_data             (SpiceVDAgent *agent, guint8 selection,
                                         const gchar *type, gpointer data, gsize len);
void vdagent_clipboard_release          (SpiceVDAgent *agent, guint8 selection);
void vdagent_clipboard_release_all      (SpiceVDAgent *agent);

G_END_DECLS

#endif


src/vdagent/vdagent.c
/*
 * vdagent session
 * Copyright (C) 2014  Red Hat, Inc.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include <sys/select.h>
#include <sys/stat.h>

#include "utils.h"
#include "vdagent.h"
#include "vdagent-clipboard.h"

G_DEFINE_TYPE (SpiceVDAgent, spice_vdagent, G_TYPE_OBJECT);

static const char *vdagentd_socket = "/var/run/spice-vdagentd/spice-vdagent-sock";
static gboolean version_mismatch = FALSE;
static gboolean quit = FALSE;

static void read_new_message(SpiceVDAgent *agent);
static void send_xorg_config(SpiceVDAgent *self);

static void
spice_vdagent_init(SpiceVDAgent *self)
{
    self->cancellable = g_cancellable_new();

    vdagent_clipboard_init(self);

    self->outq = g_queue_new();

    g_signal_connect_swapped(gdk_screen_get_default(), "monitors-changed",
                             G_CALLBACK(send_xorg_config), self);
}

static void
spice_vdagent_finalize(GObject *gobject)
{
    SpiceVDAgent *self = SPICE_VDAGENT(gobject);

    g_signal_handlers_disconnect_by_data(gdk_screen_get_default(), self);

    g_clear_object(&self->connection);
#ifdef G_OS_UNIX
    g_clear_object(&self->connectable);
    g_clear_object(&self->socket);
#endif

    g_free(self->data);
    g_queue_free_full(self->outq, g_free);

    if (G_OBJECT_CLASS(spice_vdagent_parent_class)->finalize)
        G_OBJECT_CLASS(spice_vdagent_parent_class)->finalize(gobject);
}

static void spice_vdagent_class_init(SpiceVDAgentClass *klass)
{
    GObjectClass *gobject_class = G_OBJECT_CLASS(klass);

    gobject_class->finalize = spice_vdagent_finalize;
}

typedef struct _Msg {
    gsize size;
    guint8 *data;
    GFreeFunc free_func;
} Msg;

static void kick_write(SpiceVDAgent *self);

static void
msg_write_cb(GObject *source_object,
             GAsyncResult *res,
             gpointer user_data)
{
    SpiceVDAgent *self = SPICE_VDAGENT(user_data);
    GError *error = NULL;
    Msg *msg;
    gssize ret;

    msg = g_queue_peek_head(self->outq);
    g_return_if_fail(msg != NULL);

    ret = g_output_stream_write_finish(G_OUTPUT_STREAM(source_object), res, &error);
    self->writing = FALSE;

    if (ret == -1) {
        g_warning("failed to write: %s", error->message);
        g_clear_error(&error);
    } else {
        self->pos += ret;
        g_assert(self->pos <= msg->size);

        g_debug("wrote %" G_GSSIZE_FORMAT "/%" G_GSSIZE_FORMAT, self->pos,  msg->size);
        if (self->pos == msg->size) {
            if (msg->free_func)
                msg->free_func(msg->data);
            g_slice_free(Msg, g_queue_pop_head(self->outq));
            self->pos = 0;
        }
        kick_write(self);
    }
}

static void
kick_write(SpiceVDAgent *self)
{
    Msg *msg = g_queue_peek_head(self->outq);

    if (!msg || self->writing)
        return;

    GOutputStream *out = g_io_stream_get_output_stream(self->connection);
    g_output_stream_write_async(out, msg->data + self->pos,
                                msg->size - self->pos,
                                G_PRIORITY_DEFAULT, self->cancellable,
                                msg_write_cb, self);
    self->writing = TRUE;
}

void
spice_vdagent_write(SpiceVDAgent *self,
                    gpointer data, guint32 size,
                    GFreeFunc free_func)
{
    g_return_if_fail(SPICE_IS_VDAGENT(self));
    g_return_if_fail(data && size);

    Msg *msg = g_slice_new(Msg);

    msg->size = size;
    msg->data = data;
    msg->free_func = free_func;

    g_queue_push_tail(self->outq, msg);
    kick_write(self);
}

void
spice_vdagent_write_header(SpiceVDAgent *self,
                           guint32 type, guint32 arg1, guint32 arg2, guint32 size)
{
    g_return_if_fail(SPICE_IS_VDAGENT(self));

    VDAgentdHeader *header = g_new(VDAgentdHeader, 1);
    header->type = type;
    header->arg1 = arg1;
    header->arg2 = arg2;
    header->size = size;

    spice_vdagent_write(self, header, sizeof(*header), g_free);
}

void
spice_vdagent_write_msg(SpiceVDAgent *self,
                        guint32 type, guint32 arg1, guint32 arg2,
                        gpointer data, guint32 size, GFreeFunc free_func)
{

    g_return_if_fail(SPICE_IS_VDAGENT(self));

    spice_vdagent_write_header(self, type, arg1, arg2, size);

    if (size)
        spice_vdagent_write(self, data, size, free_func);
}

typedef struct _VDAgentdRes {
    int width;
    int height;
    int x;
    int y;
} VDAgentdRes;

static void
send_xorg_config(SpiceVDAgent *self)
{
    GdkRectangle mon;
    int i, nres = gdk_screen_get_n_monitors(gdk_screen_get_default());
    VDAgentdRes *res = g_new(VDAgentdRes, nres);

    for (i = 0; i < nres; i++) {
        gdk_screen_get_monitor_geometry(gdk_screen_get_default(), i, &mon);
        res[i].width = mon.width;
        res[i].height = mon.height;
        res[i].x = mon.x;
        res[i].y = mon.y;
    }

    spice_vdagent_write_msg(self, VDAGENTD_GUEST_XORG_RESOLUTION,
                            gdk_screen_width(), gdk_screen_height(),
                            res, sizeof(*res) * nres, g_free);
}

static void
dispatch_message(SpiceVDAgent *agent)
{
    VDAgentdHeader *header = &agent->header;
    gpointer data = agent->data;

    switch (header->type) {
    case VDAGENTD_MONITORS_CONFIG:
        g_warning("monitors config is not handled by agent anymore");
        break;
    case VDAGENTD_FILE_XFER_START:
    case VDAGENTD_FILE_XFER_STATUS:
    case VDAGENTD_FILE_XFER_DATA:
        g_warning("file-xfer is deprecated");
        break;
    case VDAGENTD_CLIPBOARD_REQUEST:
        vdagent_clipboard_request(agent, header->arg1, header->arg2);
        break;
    case VDAGENTD_CLIPBOARD_GRAB:
        vdagent_clipboard_grab(agent, header->arg1,
                               (guint32*)data, header->size / sizeof(guint32));
        break;
    case VDAGENTD_CLIPBOARD_DATA:
        vdagent_clipboard_data(agent, header->arg1, header->arg2,
                             data, header->size);
        break;
    case VDAGENTD_CLIPBOARD_RELEASE:
        vdagent_clipboard_release(agent, header->arg1);
        break;
    case VDAGENTD_VERSION:
        if (g_strcmp0(data, VERSION)) {
            g_message("vdagentd version mismatch: got %s expected %s", (gchar*)data, VERSION);
            gtk_main_quit();
            version_mismatch = TRUE;
        }
        break;
    case VDAGENTD_CLIENT_DISCONNECTED:
        vdagent_clipboard_release_all(agent);
        break;
    default:
        g_warning("Unknown message from vdagentd type: %d, ignoring", header->type);
    }
}

static void
message_data_cb(GObject *source_object,
                GAsyncResult *res,
                gpointer user_data)
{
    SpiceVDAgent *agent = user_data;
    VDAgentdHeader *header = &agent->header;
    GError *error = NULL;
    gboolean success;
    gsize bread;

    success = input_stream_read_all_finish(G_INPUT_STREAM(source_object), res, &bread, &error);
    if (!success)
        g_warning("%s", error->message);

    if (bread != header->size) {
        g_warning("failed to read all data, quit: %" G_GSIZE_FORMAT, bread);
        gtk_main_quit();
    } else {
        dispatch_message(agent);
        read_new_message(agent);
    }

    g_clear_error(&error);
}

static void
message_header_cb(GObject *source_object,
                  GAsyncResult *res,
                  gpointer user_data)
{
    SpiceVDAgent *agent = user_data;
    VDAgentdHeader *header = &agent->header;
    GError *error = NULL;
    gboolean success;
    gsize bread;

    success = input_stream_read_all_finish (G_INPUT_STREAM(source_object), res, &bread, &error);
    if (!success)
        g_warning("%s", error->message);

    if (bread != sizeof(*header)) {
        g_warning("failed to read all header, quit");
        gtk_main_quit();
    } else {
        g_debug("Header type:%u size:%u (%u, %u)",
                header->type, header->size, header->arg1, header->arg2);

        if (header->size != 0) {
            agent->data = g_realloc(agent->data, header->size);
            input_stream_read_all_async(g_io_stream_get_input_stream(agent->connection),
                                        agent->data, header->size,
                                        G_PRIORITY_DEFAULT,
                                        agent->cancellable,
                                        message_data_cb,
                                        agent);
        } else {
            dispatch_message(agent);
            read_new_message(agent);
        }
    }

    g_clear_error(&error);
}

static void
read_new_message(SpiceVDAgent *agent)
{
    input_stream_read_all_async(g_io_stream_get_input_stream(agent->connection),
                                &agent->header, sizeof(agent->header),
                                G_PRIORITY_DEFAULT,
                                agent->cancellable,
                                message_header_cb,
                                agent);
}

/* TODO: could use dbus, but for portability maybe not..... */
static void
do_connect(SpiceVDAgent *agent)
{
    GError *error = NULL;
    GSocketAddressEnumerator *enumerator;
    GSocketAddress *address;

#ifdef G_OS_UNIX
    agent->socket = g_socket_new(G_SOCKET_FAMILY_UNIX, G_SOCKET_TYPE_STREAM, 0, &error);
    agent->connectable = G_SOCKET_CONNECTABLE(g_unix_socket_address_new(vdagentd_socket));
#endif

    enumerator = g_socket_connectable_enumerate (agent->connectable);
    while (TRUE) {
        address = g_socket_address_enumerator_next (enumerator, agent->cancellable, NULL);
        if (!address) {
            g_error("can't connect to socket");
            exit(1);
        }

        if (g_socket_connect(agent->socket, address, agent->cancellable, &error))
            break;

        g_debug("Connection failed, trying next");
        g_clear_error (&error);
        g_object_unref (address);
    }
    g_object_unref(enumerator);

    agent->connection =
        G_IO_STREAM(g_socket_connection_factory_create_connection(agent->socket));
    g_debug("Connected to %s %p", vdagentd_socket, agent->connection);

    send_xorg_config(agent);
}

static void
quit_handler(int sig)
{
    quit = TRUE;
    gtk_main_quit();
}

static void
run_agent(void)
{
    SpiceVDAgent *agent = g_object_new(SPICE_TYPE_VDAGENT, NULL);

    g_debug("Starting up new session");

    do_connect(agent);
    read_new_message(agent);

    gtk_main();

    g_debug("Finishing session");
    g_clear_object(&agent);
}

static GOptionEntry options[] =
{
  { "socket", 'S', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_FILENAME, &vdagentd_socket, "vdagentd socket", "PATH" },
  { NULL }
};

int
main(int argc, char *argv[])
{
    GError *error = NULL;
    GOptionContext *context;

    context = g_option_context_new("- spice-vdagent " VERSION);
    g_option_context_add_main_entries(context, options, NULL);
    g_option_context_add_group(context, gtk_get_option_group (TRUE));
    if (!g_option_context_parse(context, &argc, &argv, &error)) {
        g_print("option parsing failed: %s\n", error->message);
        exit(1);
    }

    gsyslog_install("spice-vdagent");

    {
        struct sigaction act = { .sa_flags = SA_RESTART, .sa_handler = quit_handler };
        sigaction(SIGINT, &act, NULL);
        sigaction(SIGTERM, &act, NULL);
        sigaction(SIGQUIT, &act, NULL);
    }

    while (!quit) {
        if (version_mismatch) {
            g_warning("Version mismatch, restarting");
            sleep(1);
            execvp(argv[0], argv);
        }

        run_agent();
    }

    return 0;
}
